### PR TITLE
Fix/posix timespec failure on microchip

### PR DESCRIPTION
**Pull request by aggarw13**
Fix failing `UTILS_TimespecAdd` test on Microchip Curiosity PIC32MZEF board. 

Description
-----------
**Cause of failure**
POSIX utils depend on the signed integer definition of `time_t` type (used for `struct timespec.tv_sec` member), but the latest Microchip toolchains (`v2.40`+) changed the underlying type of `time_t` to be an unsigned integer

**Fix Changes**
- Update `UTILS_TimespecAdd` function to support both signed and unsigned definitions of `time_t` (Some vendor toolchains are using the unsigned type to delay the **2038** problem for 32 bit systems)
- Update the failing unit test to be separated into its own _overflow_ named test case. Update the test to use different values for signed and unsigned definitions of `time_t` to test overflow depending on platform definition of `time_t`

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 1]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
 
-        /* check for overflow in case nsec value was invalid */
-        if( pxResult->tv_nsec < 0 )
+        configPRINTF( ( "Values after nsec addition: tv_nsec=%ld, tv_sec=%d\n",
+                        pxResult->tv_nsec, pxResult->tv_sec ) );
+
+        /* Check if the addition resulted in an overflow.*/
+        if( ( pxResult->tv_nsec < x->tv_nsec ) || ( pxResult->tv_nsec < y->tv_nsec ) )
         {
+            configPRINTF( ( "Detected overflow in tv_nsec: ReturnStatus=%d", iStatus ) );
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:217]: This will print incorrect ReturnStatus. `iStatus` needs to be sets first before printing.

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 2]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
+    /* Check input parameters. None of the parameters should be NULL,
+     * and the input parameters values should be positive. */
+    if( ( pxResult == NULL ) || ( x == NULL ) || ( y == NULL ) ||
+        ( x->tv_sec < 0 ) || ( x->tv_nsec < 0 ) ||
+        ( y->tv_sec < 0 ) || ( y->tv_nsec < 0 ) )
     {
         iStatus = -1;
     }
 
+    configPRINTF( ( "Input tv_sec values: x->tv_sec: %u, y->tv_sec:%u",
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:203]: This `configPRINTF` call uses `%u` for tv_sec, while the next one uses `%d`. Please use same across the function. 

Also, some platforms may generate warnings for mismatching printf specifier (eg, %u used for signed). It is probably better to use an explicit cast:

```c
configPRINTF( ( "Input tv_sec values: x->tv_sec: %u, y->tv_sec:%u", ( uint32_t ) x->tv_sec, ( uint32_t ) y->tv_sec ) );
```

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 3]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
+
     if( iStatus == 0 )
     {
-        /* Perform addition. */
+        /* Perform addition for nanoseconds result. */
         pxResult->tv_nsec = x->tv_nsec + y->tv_nsec;
 
-        /* check for overflow in case nsec value was invalid */
-        if( pxResult->tv_nsec < 0 )
+        configPRINTF( ( "Values after nsec addition: tv_nsec=%ld, tv_sec=%d\n",
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:211]: `pxResult->tv_sec` is not initialed yet. There is no point in printing it.

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 4]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
         iStatus = -1;
     }
 
+    configPRINTF( ( "Input tv_sec values: x->tv_sec: %u, y->tv_sec:%u",
+                    x->tv_sec, y->tv_sec ) );
+
     if( iStatus == 0 )
     {
-        /* Perform addition. */
+        /* Perform addition for nanoseconds result. */
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:208]: for nanoseconds result. -->  for nanoseconds members.

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 5]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
         pxResult->tv_nsec = x->tv_nsec + y->tv_nsec;
 
-        /* check for overflow in case nsec value was invalid */
-        if( pxResult->tv_nsec < 0 )
+        configPRINTF( ( "Values after nsec addition: tv_nsec=%ld, tv_sec=%d\n",
+                        pxResult->tv_nsec, pxResult->tv_sec ) );
+
+        /* Check if the addition resulted in an overflow.*/
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:214]: Space after period.

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 6]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
-            /* check for overflow */
-            if( pxResult->tv_sec < 0 )
+            configPRINTF( ( "Values after sec addition: tv_nsec=%ld, tv_sec=%d\n",
+                            pxResult->tv_nsec, pxResult->tv_sec ) );
+
+            /* Check if addition resulted in an overflow. */
+            if( ( pxResult->tv_sec < x->tv_sec ) || ( pxResult->tv_sec < y->tv_sec ) ||
+                ( pxResult->tv_sec < llPartialSec ) )
             {
+                configPRINTF( ( "Detected overflow in tv_sec: ReturnStatus=%d", iStatus ) );
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:236]: Same as above. `configPRINTF` needs to move after `iStatus = 1;`.

File: vendors/infineon/boards/xmc4800_iotkit/aws_tests/config_files/iot_config.h [hunk 7]
```diff
@@ -27,6 +27,10 @@
 /* Platform thread priority. */
 #define IOT_THREAD_DEFAULT_PRIORITY      5
 
+#define IOT_LOG_LEVEL_HTTPS              IOT_LOG_DEBUG
+
+#define IOT_THREAD_DEFAULT_STACK_SIZE    4096
```
Comment (aggarg) [inline vendors/infineon/boards/xmc4800_iotkit/aws_tests/config_files/iot_config.h:32]: This does not seem related to timespec.

File: vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_tests/config_files/iot_config.h [hunk 8]
```diff
@@ -1,31 +1,33 @@
...
 /* This file contains configuration settings for the demos. */
 
 #ifndef IOT_CONFIG_H_
 #define IOT_CONFIG_H_
 
 /* Platform thread priority. */
-#define IOT_THREAD_DEFAULT_PRIORITY      5
+#define IOT_THREAD_DEFAULT_PRIORITY    5
+
+#define IOT_LOG_LEVEL_HTTPS            IOT_LOG_DEBUG
```
Comment (aggarg) [inline vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_tests/config_files/iot_config.h:30]: Does not seem related to timespec.

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c [hunk 9]
```diff
@@ -84,6 +97,31 @@ TEST( Full_POSIX_UTILS, UTILS_TimespecAdd )
...
+        x.tv_sec = -1;
+        TEST_ASSERT_EQUAL_INT( -1, UTILS_TimespecAdd( &x, &y, NULL ) );
+
+        x.tv_sec = 0;
+        y.tv_sec = -2;
+        TEST_ASSERT_EQUAL_INT( -1, UTILS_TimespecAdd( &x, &y, NULL ) );
+    }
+
+    /* Test with invalid parameter check when tv_nsec < 0. */
+    if( isTimespecTvSecSigned == 1u )
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:112]: Can be in the same if check as above.

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c [hunk 10]
```diff
@@ -44,14 +44,26 @@
 #include "unity_fixture.h"
 #include "unity.h"
 
+/*-----------------------------------------------------------*/
+/* Constants that represent max values for signed and unsigned definitions of time_t */
+#define TIME_T_SIGNED_MAX      ( time_t ) ( ( ( time_t ) 1 << ( ( sizeof( time_t ) << 3 ) - 1 ) ) - 1 )
+#define TIME_T_UNSIGNED_MAX    ( ( time_t ) ULLONG_MAX )
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:50]: ~0 is better instead of assuming unsigned long long.

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c [hunk 11]
```diff
@@ -44,14 +44,26 @@
 #include "unity_fixture.h"
 #include "unity.h"
 
+/*-----------------------------------------------------------*/
+/* Constants that represent max values for signed and unsigned definitions of time_t */
+#define TIME_T_SIGNED_MAX      ( time_t ) ( ( ( time_t ) 1 << ( ( sizeof( time_t ) << 3 ) - 1 ) ) - 1 )
```
Comment (aggarg) [inline libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:49]: Why to assume that tv_sec is time_t? Is it possible to use `sizeof(signedCheck.tv_sec)` ?

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 1 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:217]: These log statements are only for debugging. Will remove them once I have a successful test execution result for Microchip on CI

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 12]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
                        struct timespec * const pxResult )
 {
-    int64_t llPartialSec = 0;
+    uint64_t llPartialSec = 0;
     int iStatus = 0;
-
-    /* Check parameters. */
-    if( ( pxResult == NULL ) || ( x == NULL ) || ( y == NULL ) )
+    struct timespec temp = { .tv_sec = 1, .tv_nsec = 1 };
+    uint8_t isSigned = 1u;
```
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:192]: Not anymore. Removed

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 13]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
                        const struct timespec * const y,
                        struct timespec * const pxResult )
 {
-    int64_t llPartialSec = 0;
+    uint64_t llPartialSec = 0;
     int iStatus = 0;
-
-    /* Check parameters. */
-    if( ( pxResult == NULL ) || ( x == NULL ) || ( y == NULL ) )
+    struct timespec temp = { .tv_sec = 1, .tv_nsec = 1 };
```
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:191]: Removed this local as is not used anymore

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 14]
```diff
@@ -186,34 +186,55 @@ int UTILS_TimespecAdd( const struct timespec * const x,
                        const struct timespec * const y,
                        struct timespec * const pxResult )
 {
-    int64_t llPartialSec = 0;
+    uint64_t llPartialSec = 0;
```
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:189]: As this variable will only be used when the `tv_nsec` addition does not overflow, this is guaranteed to be positive; thus, changed the type. 
Will update the name as well.

File: libraries/abstractions/posix/include/FreeRTOS_POSIX/utils.h [hunk 15]
```diff
@@ -93,11 +93,15 @@ void UTILS_NanosecondsToTimespec( int64_t llSource,
 /**
  * @brief Calculates pxResult = x + y.
  *
+ * This function supports both signed and unsigned integer definitions of
+ * the time_t type that is used for the timespec.tv_sec member.
```
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/abstractions/posix/include/FreeRTOS_POSIX/utils.h]: Good point; updated.

File: libraries/abstractions/posix/include/FreeRTOS_POSIX/utils.h [hunk 16]
```diff
@@ -93,11 +93,15 @@ void UTILS_NanosecondsToTimespec( int64_t llSource,
...
+ *
  * @param[in] x The first argument for addition.
  * @param[in] y The second argument for addition.
  * @param[out] pxResult Where the result of the calculation is stored.
  *
- * @return -1 if any argument was NULL; 1 if result is negative (overflow); otherwise, 0.
+ * @return -1 if any argument was NULL or any of the input values is negative;
```
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/abstractions/posix/include/FreeRTOS_POSIX/utils.h]: Fixed

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 4 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c]: Updated

File: vendors/infineon/boards/xmc4800_iotkit/aws_tests/config_files/iot_config.h (hunk 7 above)
Comment (aggarw13) [reply vendors/infineon/boards/xmc4800_iotkit/aws_tests/config_files/iot_config.h]: Removed

File: vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_tests/config_files/iot_config.h (hunk 8 above)
Comment (aggarw13) [reply vendors/infineon/boards/xmc4800_plus_optiga_trust_x/aws_tests/config_files/iot_config.h]: Removed

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c (hunk 11 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:49]: `sizeof( signedCheck.tv_sec)` would work but the `time_t` cast for the `1`  numeric constant needs to be present. I thought of using `typeof` (as `( typeof(signedCheck.tv_sec) ) ( ( ( typeof(signedCheck.tv_sec) ) 1 << ( ( sizeof( signedCheck.tv_sec ) << 3 ) - 1 ) ) - 1 )`  ) but that utility is not defined by the ISO C standard: http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 5 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c]: Fixed

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c (hunk 9 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:112]: Updated this check to instead use the `isTimespecTvNsecSigned` global variable to pass negative values only when the `tv_nsec` member type is signed

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (cobusve) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c]: updating the name of a member of a public struct will break the interface contract. There should be a good reason to do that, which I guess is why Gaurav was asking if it is really necessary to break the contract like this?

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c [hunk 17]
```diff
@@ -186,33 +186,40 @@ int UTILS_TimespecAdd( const struct timespec * const x,
...
+    uint64_t ullPartialSec = 0;
     int iStatus = 0;
 
-    /* Check parameters. */
-    if( ( pxResult == NULL ) || ( x == NULL ) || ( y == NULL ) )
+    /* Check input parameters. None of the parameters should be NULL,
+     * and the input parameters values should be positive. */
+    if( ( pxResult == NULL ) || ( x == NULL ) || ( y == NULL ) ||
+        ( x->tv_sec < 0 ) || ( x->tv_nsec < 0 ) ||
+        ( y->tv_sec < 0 ) || ( y->tv_nsec < 0 ) )
```
Comment (cobusve) [inline libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:196]: Is this degenerate unsigned comparison not generating a compiler warning when tv_sec is unsigned?

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c (hunk 11 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:49]: Also, have verified that a number of boards (Windows, NXP, Infineon )do not support the `typeof` utility.

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 17 above)
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c:196]: How will we check against negative values otherwise?
I think that the following implementation may also generate warnings:
```
struct timespec temp;
temp.tv_sec = 1;
temp.tv_nsec = 1;

uint8_t isTvSecSigned = (~temp.tv_sec < 0);
uint8_t isTvNsecSigned = (~temp.tv_nsec < 0);
if( ( (isTvSecSigned == 1) && ( ( x->tv_sec < 0) || ( y->tv_sec < 0) ) )  ||
    ( (isTvNSecSigned == 1) && ( ( x->tv_nsec < 0) || ( y->tv_nsec < 0) ) ) )
{
 iStatus = -1;
}
```

File: libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/source/FreeRTOS_POSIX_utils.c]: This is not a member of the public struct, is it?

File: libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c (hunk 11 above)
Comment (cobusve) [reply libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:49]: I think the purpose of this is to contain the max value of TIME_T and not the max value of the structure member, if the intent is the latter the name should reflect what it is.

I am not convinced that the logic is going to work all the time here. Lets test with time_t = 32 bit int.

It looks like you are doing a left-shift of 1 by 31 (which makes the intermediate result a signed int with most significant bit set, or a negative number) and then subtracting one. This seems dangerous to do, I have not tested to check the behavior but I would to know that you can explain why this is safe to do here?

Comment (aggarw13) [reply libraries/freertos_plus/standard/freertos_plus_posix/test/iot_test_posix_utils.c:49]: This logic has been taken from the Linux kernel implementation (`v4.7`), https://elixir.bootlin.com/linux/v4.7/source/include/linux/time.h#L11

I have tested that all the boards run successfully for this test.
